### PR TITLE
Add PHP 8.4

### LINKED ISSUE #37523: PHP 8.4 compatibility

### LINKED ISSUE #37528: Ensure that all dependencies are also compatible

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Provides partial support for PHP 8.4.
| Type?             | refacto
| Category?         | PM
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | Auto test
| UI Tests          | PHP 8.1: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/12090483083<br/>PHP 8.2: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/12090487636<br/>PHP 8.3: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/12090493576<br/>PHP 8.4: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/12089618004
| Fixed issue or discussion?     | Fixes #37528 and Fixes #37523
| Related PRs       | https://github.com/PrestaShop/docker/pull/406
| Sponsor company   | 

### Composer commands run

- composer update friendsofsymfony/jsrouting-bundle (^3.5)
- composer update mikey179/vfsstream
- composer update twig/extra-bundle twig/string-extra
- composer update monolog/monolog (^2)
- composer update bjeavons/zxcvbn-php (^1.4)
- composer update behat/behat -W (^3.15)
- composer update greenlion/php-sql-parser(^4.7 based on PrestaShop fork https://github.com/PrestaShop/PHP-SQL-Parser)
- composer update ergebnis/composer-normalize -W (^2.44)
- composer update doctrine/orm -W (^2.15)
- composer update prestashop/ps_linklist
- composer update phpstan/phpstan phpstan/phpstan-doctrine

### Partial PRs
- https://github.com/PrestaShop/PrestaShop/pull/37414
- https://github.com/PrestaShop/PrestaShop/pull/37415
- https://github.com/PrestaShop/PrestaShop/pull/37416
- https://github.com/PrestaShop/PrestaShop/pull/37434
- https://github.com/PrestaShop/PrestaShop/pull/37429
- https://github.com/PrestaShop/PrestaShop/pull/37463
- https://github.com/PrestaShop/PrestaShop/pull/37515
- https://github.com/PrestaShop/ps_linklist/pull/204 & https://github.com/PrestaShop/ps_linklist/pull/205
- https://github.com/PrestaShop/ps_facetedsearch/pull/1094 & https://github.com/PrestaShop/ps_facetedsearch/pull/1107
- https://github.com/PrestaShop/autoupgrade/pull/1039 & release
- https://github.com/PrestaShop/ps_emailalerts/pull/146 & https://github.com/PrestaShop/ps_emailalerts/pull/148

### CODE REVIEW

File: tests/Unit/Classes/ToolsTest.php [hunk 1]
```diff
@@ -474,7 +478,7 @@ public function providerMathRound(): array
      */
     public function testMathRound(float $expectedResult, float $value, int $precision, int $mode): void
     {
-        $this->assertSame($expectedResult, Tools::math_round($value, $precision, $mode));
+        $this->assertSame($expectedResult, Tools::ps_round($value, $precision, $mode));
```
Comment (jolelievre) [inline tests/Unit/Classes/ToolsTest.php]: Why change this, the test is supposed to test the `math_round` method not the `ps_round` one

File: src/PrestaShopBundle/Service/Log/LogHandler.php [hunk 2]
```diff
@@ -50,13 +44,15 @@ public function __construct(Container $container, $level = Logger::DEBUG, $bubbl
 
     /**
      * Writes the record down to the log of the implementing handler
-     *
-     * @phpstan-param FormattedRecord $record
      */
-    protected function write(array $record): void
+    protected function write(LogRecord $record): void
```
Comment (jolelievre) [inline src/PrestaShopBundle/Service/Log/LogHandler.php]: Was the update of monolog that forces this modification a requirement?

Comment (Hlavtox) [reply src/PrestaShopBundle/Service/Log/LogHandler.php]: Yes, see https://github.com/Seldaek/monolog/blob/main/UPGRADE.md

**Comment by kpodemski** [comment]
tests green 🥹

**Review by jolelievre — APPROVED** [review]
Thanks @Hlavtox

**Review by boherm — APPROVED** [review]

**Pull request merged**